### PR TITLE
Add Latest Posts block to email editor

### PR DESCRIPTION
**Pull request by yuliyan**
## Description

Adds a **Latest Posts** block to the block-based email editor. When the email is sent, it automatically pulls in your most recent posts and renders them with a per-post layout you design once and it repeats.

It uses two cooperating blocks - a similar container + repeating-template pattern as core's Query Loop / Post Template:

- **`mailpoet/latest-posts`** (container) - selection control - count, post type, order, offset, category/tag filtering (include or exclude), manual post selection, and 1–2 column layout.
- **`mailpoet/latest-posts-template`** (template) - the inner-block layout for a single post. It has a default layout of featured image, title, excerpt, but it allows full customization - add, remove, and move blocks.

The editor preview mirrors the server-side query so what you see matches what gets sent. At render time, posts are selected dynamically and the template's inner blocks are rendered once per post, with email-safe HTML.

## Code review notes

- `LatestPosts.php` is the render-side entry point. It registers both blocks, opts core inner blocks into email rendering, and wires the `render_email_callback` the email editor expects.
- `post-template/edit.tsx` (`buildRestQuery`) mirrors the server-side `BlockPostQuery` to keep preview and sent output in sync.

## QA notes

1. Add the **Latest Posts** block in the email editor. Confirm the default layout previews recent posts.
2. Adjust selection controls (count, order, offset, columns) and confirm the preview updates.
3. Filter by category and tag, using both include and exclude.
4. Use manual selection - pick/reorder posts; confirm an empty selection shows "No posts found."
5. Customize the per-post template and confirm it repeats for every post.
6. Send a test email and confirm it matches the preview.

## Linked PRs

_N/A_

## Linked tickets

[STOMAIL-8179: Add Posts / Automatic Latest Content email editor block](https://linear.app/a8c/issue/STOMAIL-8179/add-posts-automatic-latest-content-email-editor-block)

## After-merge notes

_N/A_

## Screenshots or screencast 

| Email editor                          | Email preview                          |
| ------------------------------- | ------------------------------ |
| 4vI0G83S3pskSdLf@2x | KsH8j3O7tOO80A5G@2x |

## Tasks

- [ ] I have added a changelog entry (`pnpm changelog:add --type=<type> --description=<description>`)
- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-8179-add-posts-automatic-latest-content-email-editor-block)

_The latest successful build from `stomail-8179-add-posts-automatic-latest-content-email-editor-block` will be used. If none is available, the link won't work._

No issues found.

### CODE REVIEW

**Comment by yuliyan** [comment]
@coderabbitai review

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+      'contentType' => isset($query['postType']) && is_string($query['postType']) ? $query['postType'] : 'post',
+    ];
+
+    if ($isManual) {
+      $args['posts'] = $manualPosts;
+      return $args;
+    }
+
+    $args['amount'] = isset($query['perPage']) && is_numeric($query['perPage']) ? (int)$query['perPage'] : 3;
+    $args['offset'] = isset($query['offset']) && is_numeric($query['offset']) ? (int)$query['offset'] : 0;
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php]: Fixed in https://github.com/mailpoet/mailpoet/pull/6917/commits/8c1666c887f79a4dd0238852419f3334d57e68af.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                selectionMode: value === 'manual' ? 'manual' : 'latest',
+              })
+            }
+            __nextHasNoMarginBottom
+          />
+          <SelectControl
+            label={__('Content type', 'mailpoet')}
+            value={query.postType}
+            options={POST_TYPE_OPTIONS}
+            onChange={(value) => updateQuery({ postType: value })}
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx:238]: Fixed in https://github.com/mailpoet/mailpoet/pull/6917/commits/8c1666c887f79a4dd0238852419f3334d57e68af.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  attributes,
+  setAttributes,
+}: BlockEditProps<Attributes>): JSX.Element {
+  const { query, displayLayout } = attributes;
+  const blockProps = useBlockProps();
+  const innerBlocksProps = useInnerBlocksProps(blockProps, {
+    allowedBlocks: [TEMPLATE_BLOCK_NAME],
+    template: INNER_TEMPLATE,
+    templateLock: false,
+  });
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx:210]: Fixed in https://github.com/mailpoet/mailpoet/pull/6917/commits/8c1666c887f79a4dd0238852419f3334d57e68af.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    .map((id) => selectedPosts.find((post) => post.id === id))
+    .filter((post): post is PostRecord => Boolean(post))
+    .map(postTitle);
+
+  const handleChange = (tokens: (string | { value: string })[]): void => {
+    const ids = tokens
+      .map((token) => (typeof token === 'string' ? token : token.value))
+      .map((title) => titleToId.get(title))
+      .filter((id): id is number => typeof id === 'number');
+    onChange(ids);
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx]: Fixed in https://github.com/mailpoet/mailpoet/pull/6917/commits/8c1666c887f79a4dd0238852419f3334d57e68af.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 5]
```diff
@@ -0,0 +1,856 @@
...
+
+/**
+ * Renders the MailPoet "Latest posts" block for the email editor.
+ *
+ * There are two blocks working together here: the container (mailpoet/latest-posts)
+ * is where you pick which posts to show, and it wraps a template block
+ * (mailpoet/latest-posts-template) that defines how a single post looks. When we
+ * render, we fetch the posts and repeat the template's inner blocks for each one.
+ */
+class LatestPosts extends AbstractBlock {
```
Comment (costasovo) [inline mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:30]: This class is pretty huge, and I can imagine we will add more lines in the future. I would suggest splitting it into a few classes. Could also be helpful for adding tests.

Perhaps something like:
LatestPost - init, hooks
LatestPostFetcher - getPosts, queryArgs, getNewsletterIdForQuery, ...
LatestPostRenderer - rendering

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+   */
+  private function getNewerThanTimestamp(?NewsletterEntity $newsletter) {
+    if (!$newsletter || $newsletter->getType() !== NewsletterEntity::TYPE_NOTIFICATION_HISTORY) {
+      return false;
+    }
+    $parent = $newsletter->getParent();
+    if (!$parent instanceof NewsletterEntity) {
+      return false;
+    }
+    $lastPost = $this->newsletterPostsRepository->findOneBy(['newsletter' => $parent], ['createdAt' => 'desc']);
```
Comment (costasovo) [inline mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:1157]: I see we are already preparing for usage in post notifications.
There is one important thing about how we detect which posts were already sent. After we render the email in real sending, [we look for `data-post-id` attributes in the HTML](https://github.com/mailpoet/mailpoet/blob/88005ccca6ea500577b6f5972b0c10c6076fc84f/mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Posts.php#L31-L36).

I'm not sure we want to reuse this mechanism. Maybe initially, yes? So to make it work, we need to sneak in the `data-post-id` attribute somewhere into the post HTML.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 7]
```diff
@@ -0,0 +1,856 @@
...
+   */
+  private function buildQueryArgs(array $query, bool $isManual, array $manualPosts): array {
+    $args = [
+      'contentType' => isset($query['postType']) && is_string($query['postType']) ? $query['postType'] : 'post',
+    ];
+
+    if ($isManual) {
+      $args['posts'] = $manualPosts;
+      return $args;
+    }
```
Comment (costasovo) [inline mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:621]: An issue flagged by agent:
```
Manual selection loses pick order. Manual mode returns early with no sortBy, so getOrder() → ASC + orderby='date'; post__in does not preserve array order.
```
I was able to replicate

### Editor

Image

### Preview

Image

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 8]
```diff
@@ -0,0 +1,856 @@
...
+  private const MAX_POSTS = 100;
+  private const DEFAULT_BLOCK_GAP = '20px';
+  private const DEFAULT_CONTENT_WIDTH_PX = 600;
+
+  /** Template blocks. Not email-enabled by default, so we opt them in. */
+  private const TEMPLATE_CORE_BLOCKS = [
+    'core/post-featured-image',
+    'core/post-title',
+    'core/post-excerpt',
+    'core/post-content',
```
Comment (costasovo) [inline mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php]: AFAIK, `core/post-content` basically renders the full post body, which may possibly include many unsupported blocks. In the old renderer, we transform the post body for emails.
Do we do that somewhere (maybe I missed that) or is there a plan to add it as a follow-up?

**Review by costasovo — COMMENTED** [review]
@yuliyan Thank you for working on this! This is a huge and importat feature. 

There is a lot of code, and after some time, I was losing focus, so in this review, I submit my currnet remarks, and we will do more rounds. Claude flagged a couple of items as well and some of them are also included.

Appart the stuff mentioned in comments I noticed the MailPoet's Latest Posts leaks into the WP's post editor. I think it is true also for our Powered by MailPoet block, but that one is hidden from inserter so the issue surfaced with this block. 

Two latest posts blocks in WP's post editor. One of them is from MailPoet.
Image

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 8 above)
Comment (costasovo) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php]: Hmm, thinking about my comment ⬆️, one of the reasons we use core blocks in the email editor is that we want to be able to process posts, so the previous comment is a bit off. But maybe we will need some kind of block filtering to skip or convert unsupported blocks. This would be a follow-up task.

**Comment by hiSandog** [comment]
The Latest Posts block looks like it touches both editor controls and email rendering, so the edge cases are mostly content-state related. It would be good to test no matching posts, posts without featured images, private/password-protected posts, and a manual selection that later becomes unavailable. Email blocks often behave differently from normal editor blocks, so one rendered-email snapshot would add useful coverage.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 8 above)
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php]: I agree - ideally, we'll process the post content and render the supported blocks. However, we can't guarantee that all posts use the block structure, so implementing the email-safe filter is still necessary.

However, I noticed that even though I enforce email support for `core/post-content` in this PR, it's still not available in the inserter. I believe that's because the email editor reserves that block as the email's own body slot. This PR is already a bit too large, so it's probably better to create a follow-up issue once it's shipped. 

For the time being, I've removed `core/post-content` from the list of supported blocks https://github.com/mailpoet/mailpoet/pull/6917/commits/92dc93231cb2723abfc3e088adc47984edef268d.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 6 above)
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:1157]: Thanks for flagging it! I found out about the missing attribute while working on https://github.com/mailpoet/mailpoet/pull/6921 and was planning to add in a follow-up PR, as this PR was already getting too large, but the implementation turned out to be smaller than I thought so I added it anyway:

https://github.com/mailpoet/mailpoet/pull/6917/commits/1995213bb036eedfc8245d3c9bfbfcd6f848253b.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 7 above)
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:621]: Good catch! Fixed it in https://github.com/mailpoet/mailpoet/pull/6917/commits/862277925be58557ae397e44970a4eeb7813f26e.

I thought about using `"orderby" => "post__in"` in `BlockPostQuery`, but noticed that it will alter the current behavior in the legacy editor, so decided to go with this "manual sorting" approach.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 5 above)
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:30]: That's a good idea - would you mind if we refactor it in a separate PR?

Comment (costasovo) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:30]: Yes. Sure! Let's do it in a follow-up.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php (hunk 8 above)
Comment (costasovo) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php]: I think it is ok to remove for now. The core/post-content block is in the inserter only when the block editor is used for editing a template (wp_template post type). It represents the whole post body. I think that in the future, we will want to add it to allow sending full posts as we allow in the legacy editor. So we will need a workaround. Either somehow allow core/post-content or introduce our own mailpoet/post-content that could cover the email-safe filter.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+}): JSX.Element {
+  const records = useSelect((select) => {
+    const { getEntityRecords } = select(coreStore);
+    const restQuery = { per_page: -1, orderby: 'name', order: 'asc' };
+    const categories = (getEntityRecords('taxonomy', 'category', restQuery) ??
+      []) as unknown as TermRecord[];
+    const tags = (getEntityRecords('taxonomy', 'post_tag', restQuery) ??
+      []) as unknown as TermRecord[];
+    return [...categories, ...tags];
+  }, []);
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/edit.tsx:81]: `per_page: -1` is a supported case - see https://github.com/WordPress/gutenberg/blob/28c348acdf3e88bbd90f766e66acc6716bcf9103/packages/api-fetch/src/middlewares/fetch-all-middleware.ts#L74.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/post-template/edit.tsx [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    <div
+      {...blockPreviewProps}
+      tabIndex={0}
+      role="button"
+      onClick={handleOnClick}
+      onKeyDown={handleOnClick}
+      onKeyUp={handleOnClick}
+      style={{ display: isHidden ? 'none' : undefined }}
+    />
+  );
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/post-template/edit.tsx:169]: Updated in https://github.com/mailpoet/mailpoet/pull/6917/commits/d4d16a16127dc3f15b3c08c519205acb21180f56.

File: mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/post-template/edit.tsx [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    // include: [0] resolves to no posts, so an empty manual selection previews
+    // as "no posts" instead of falling back to the latest ones.
+    return manualPosts.length
+      ? {
+          include: manualPosts,
+          per_page: manualPosts.length,
+          orderby: 'include',
+        }
+      : { include: [0], per_page: 1 };
+  }
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/assets/js/src/mailpoet-custom-email-editor-blocks/latest-posts/post-template/edit.tsx:107]: Added a cap in https://github.com/mailpoet/mailpoet/pull/6917/commits/d4d16a16127dc3f15b3c08c519205acb21180f56.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 12]
```diff
@@ -0,0 +1,927 @@
...
+  private function getManualPostIds(array $query): array {
+    $posts = isset($query['posts']) && is_array($query['posts']) ? $query['posts'] : [];
+    $ids = [];
+    foreach ($posts as $post) {
+      if (is_numeric($post) && (int)$post > 0) {
+        $ids[] = (int)$post;
+      }
+    }
+    return $ids;
+  }
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:655]: Added the cap in https://github.com/mailpoet/mailpoet/pull/6917/commits/d4d16a16127dc3f15b3c08c519205acb21180f56.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 13]
```diff
@@ -0,0 +1,927 @@
...
+    $postType = $post instanceof \WP_Post ? $post->post_type : null; // phpcs:ignore Squiz.NamingConventions.ValidVariableName.MemberNotCamelCaps
+    if ($postType === EmailEditor::MAILPOET_EMAIL_POST_TYPE) {
+      return $allowedBlocks;
+    }
+
+    // When every block is allowed the value is `true` rather than a list. Build
+    // the full list of block names first so we have something to remove ours from.
+    if (!is_array($allowedBlocks)) {
+      $allowedBlocks = array_keys(\WP_Block_Type_Registry::get_instance()->get_all_registered());
+    }
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:194]: Handled the case in https://github.com/mailpoet/mailpoet/pull/6917/commits/d4d16a16127dc3f15b3c08c519205acb21180f56.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php [hunk 14]
```diff
@@ -0,0 +1,927 @@
...
+    return self::DEFAULT_CONTENT_WIDTH_PX;
+  }
+
+  private function parsePx(string $value): int {
+    $value = trim($value);
+    if ($value === '') {
+      return 0;
+    }
+    return (int)round((float)str_replace('px', '', $value));
+  }
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Blocks/BlockTypes/LatestPosts.php:1087]: Added handling (fallback) for non-px values in https://github.com/mailpoet/mailpoet/pull/6917/commits/d4d16a16127dc3f15b3c08c519205acb21180f56.
